### PR TITLE
Add glob matcher to Pipeline search

### PR DESCRIPTION
**Pull request by i386**
# Description

See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-rest-impl/pom.xml [hunk 1]
```diff
@@ -43,6 +43,11 @@
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>credentials</artifactId>
         </dependency>
+        <dependency>
```
Comment (vivek) [inline blueocean-rest-impl/pom.xml]: Should be moved under `dependencyManagement` in parent pom and declared here without version.

File: blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/GlobMatcher.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.apache.oro.text.GlobCompiler;
+import org.apache.oro.text.regex.MalformedPatternException;
+import org.apache.oro.text.regex.Pattern;
+import org.apache.oro.text.regex.Perl5Matcher;
+
+/**
+ * Matches glob string patters
+ * See https://en.wikipedia.org/wiki/Glob_(programming)
+ */
+public final class GlobMatcher {
```
Comment (vivek) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/GlobMatcher.java]: Better to move it in `blueocean-commons`?

File: blueocean-rest-impl/src/test/java/io/jenkins/blueocean/service/embedded/util/GlobMatcherTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package io.jenkins.blueocean.service.embedded.util;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class GlobMatcherTest {
```
Comment (vivek) [inline blueocean-rest-impl/src/test/java/io/jenkins/blueocean/service/embedded/util/GlobMatcherTest.java]: 👍

File: blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineSearch.java [hunk 4]
```diff
@@ -105,12 +106,15 @@ public String getType() {
                 }
             });
         }else{
+            GlobMatcher matcher = pipeline.contains("*") ? new GlobMatcher(pipeline) : null;
```
Comment (kzantow) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineSearch.java]: A lot of code for `Pattern.compile(("\\Q"+pipeline.replace("*", "\\E.*\\Q")+"\\E").replace("\\Q\\E",""))` ;)

File: blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/GlobMatcher.java (hunk 2 above)
Comment (i386) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/GlobMatcher.java]: then lots of plugins will be pulling in `oro` for no reason at all. I think its fine to leave here in `util` until such time its needed elsewhere.

File: blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineSearch.java (hunk 4 above)
Comment (i386) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineSearch.java]: you can debug all the regex that no one understands properly ;)

File: blueocean-rest-impl/pom.xml (hunk 1 above)
Comment (i386) [inline blueocean-rest-impl/pom.xml]: Fixed

File: blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/GlobMatcher.java (hunk 2 above)
Comment (vivek) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/GlobMatcher.java]: This being there in `dependencyManagement` won't make other plugins depend on it. Only if those plugin define dependencies on it. I think any module depends on a maven library should be kept in dependencyManagement, enforces one version. Just like how we have workflow libraries in dependencyManagement but only module needing them are `blueocean-pipeline-api-impl`. I think its better we are consistent than make such exceptions.

Comment (vivek) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/GlobMatcher.java]: ok

**Comment by vivek** [comment]
LGTM 🐝 pending CI/ATH run.

**Comment by i386** [comment]
ATH https://ci.blueocean.io/blue/organizations/jenkins/ATH-Jenkinsfile/detail/master/2411/pipeline

**Comment by i386** [comment]
Passes ATH https://ci.blueocean.io/blue/organizations/jenkins/ATH-Jenkinsfile/detail/master/2411/pipeline

**Comment by michaelneale** [comment]
shouldn't this go on the fast search branch?

**Comment by i386** [comment]
Nicu has merged this with his branch

**Pull request closed without merging**